File: Day21/delete-characters-to-make-fancy-string.cc
class Solution {
// Time: O(N) :: Space: O(N)
public:
    string makeFancyString(string s) {
        string newStr = "";
        char prev = '\0';
        int frequency = 0;

        for (char ch : s) {
            if (ch == prev) {
                frequency++;
            } else {
                prev = ch;
                frequency = 1;
            }

            if (frequency < 3) {
                newStr += ch;
            }
        }

        return newStr;
    }
};


class Solution {
// Time: O(N) :: Space: O(1)
public:
    string makeFancyString(string s) {
        int j = 2;
        int n = s.length();
        if (n < 3) return s;

        for (int i = 2; i < n; ++i) {
            if (!(s[i] == s[j-1] && s[i] == s[j-2])) {
                s[j++] = s[i];
            }
        }

        return s.substr(0, j);
    }
};

File: Day21/delete-characters-to-make-fancy-string.py
class Solution:
    # Time: O(N) :: Space: O(N)
    def makeFancyString(self, s: str) -> str:
        prev = ''
        frequency = 0
        newStr = ''
        for i, ch in enumerate(s):
            if ch == prev:
                frequency+=1
            else:
                prev = s[i]
                frequency = 1
            if frequency < 3:
                newStr+=ch
        return newStr


class Solution:
    # Time: O(N) :: Space: O(1)
    def makeFancyString(self, s: str) -> str:
        s = list(s)
        j = 2
        for i  in range(2, len(s)):
            if s[i] != s[j-1] or s[i] != s[j-2]:
                s[j] = s[i]
                j+=1
             
        
        return "".join(s[:j])

        